### PR TITLE
Eliminated heap

### PR DESCRIPTION
**Pull request by serges147**
- Events now stay on stack; move construct-able.

Note that events are not (yet) move assignable - possible to address, but will see do we need it actually. Implementation of such feature will make current design a bit more complex (b/c lambdas are not move assignable as soon as there is non-empty capture list (it does not matter, by value, by reference, or even dummy `[&]` without actual referencing of anything) - only explicitly empty `[]` guarantees that lambda is move assignable.

### CODE REVIEW

File: include/olg_scheduler/scheduler.hpp [hunk 1]
```diff
@@ -109,17 +98,35 @@ class Event : private cavl::Node<Event<TimePoint>>
...
+    using Tree = cavl::Tree<Event>;
+    using cavl::Node<Event>::remove;
+
     /// The event is not automatically scheduled, it must be explicitly scheduled by the user.
-    explicit Event(cavl::Tree<Event>& tree) : tree_(tree) {}
+    Event() = default;
+
+    /// The event can be safely destroyed at any time.
+    /// It will be automatically canceled and removed from the event loop.
+    virtual ~Event()
```
Comment (serges147) [inline include/olg_scheduler/scheduler.hpp:109]: GCC for some reason is not happy if I remove `virtual` from `~Event` - looks like false positive.

File: include/olg_scheduler/scheduler.hpp [hunk 2]
```diff
@@ -186,52 +188,49 @@ class EventLoop final
...
         public:
-            Impl(EventLoop& owner, const duration per, const Fun& fun) : EventProxy(owner), period_(per), handler_(fun)
+            Impl(EventLoop& owner, const duration per, Fun&& fun) : period_{per}, handler_{std::move(fun)}
             {
-                this->schedule(Clock::now() + period_);
+                this->schedule(Clock::now() + period_, owner.tree_);
             }
-            void execute(const Arg<time_point>& args) final
+
+        private:
```
Comment (serges147) [inline include/olg_scheduler/scheduler.hpp:222]: We now expose `Impl` type to user, so we have to hide private stuff, like `execute`.

File: include/olg_scheduler/scheduler.hpp [hunk 3]
```diff
@@ -135,17 +142,12 @@ class Event : private cavl::Node<Event<TimePoint>>
     /// (otherwise it will keep firing in a tight loop).
     /// If the event needs to be re-scheduled automatically, it must be done before the user callback is invoked,
     /// because the user may choose to cancel the event from the callback.
-    virtual void execute(const Arg<TimePoint>& args) = 0;
+    virtual void execute(const Arg<TimePoint>& args, Tree& tree) = 0;
 
 private:
-    cavl::Tree<Event>&       tree_;
```
Comment (serges147) [inline include/olg_scheduler/scheduler.hpp:141]: B/c of latest CAVL, we don't need anymore `tree_` for the `cancel()`. So, now even less footprint of an event.

File: include/olg_scheduler/scheduler.hpp
Comment (pavel-kirienko) [inline include/olg_scheduler/scheduler.hpp:1]: Maybe we should call this file `include/olg_scheduler/olg_scheduler.hpp` then? I am pretty confident this little library is going to stay single-header only.

File: include/olg_scheduler/scheduler.hpp [hunk 4]
```diff
@@ -18,12 +18,12 @@
 #pragma once
 
 #include <cavl/cavl.hpp>
```
Comment (pavel-kirienko) [inline include/olg_scheduler/scheduler.hpp]: The idiomatic inclusion pattern for cavl is just cavl.hpp. The current form will break if the user decided to submodule the cavl repo. The include path settings will need to be adjusted.

```suggestion
#include <cavl.hpp>
```

**Review by pavel-kirienko — APPROVED** [review]
The readme should mention that `cavl.hpp` is the only required dependency.

**Comment by pavel-kirienko** [comment]
Please transfer the repository to Zubax afterward.

**Pull request merged**